crates/task-impls/src/quorum_proposal.rs

```diff
@@ -0,0 +1,574 @@
+    /// Create an event dependency
+    #[instrument(skip_all, fields(id = self.id, latest_proposed_view = *self.latest_proposed_view), name = "Quorum proposal create event dependency", level = "error")]
+    fn create_event_dependency(
+        &self,
+        dependency_type: ProposalDependency,
+        view_number: TYPES::Time,
+        event_receiver: Receiver<Arc<HotShotEvent<TYPES>>>,
+    ) -> EventDependency<Arc<HotShotEvent<TYPES>>> {
+        EventDependency::new(
+            event_receiver,
+            Box::new(move |event| {
+                let event = event.as_ref();
+                let event_view = match dependency_type {
+                    ProposalDependency::ViewSync => {
+                        if let HotShotEvent::QuorumProposalViewSyncFinalizeCertValidated(
+                            view_sync_cert,
+                        ) = event
+                        {
+                            view_sync_cert.view_number
+                        } else {
+                            return false;
+                        }
+                    }
+                    ProposalDependency::QCFormed => {
+                        if let HotShotEvent::QuorumProposalTimeoutCertValidated(timeout_cert) =
+                            event
+                        {
+                            timeout_cert.view_number
+                        } else if let HotShotEvent::QuorumProposalQuorumCertValidated(qc) = event {
+                            qc.view_number
+                        } else {
+                            return false;
+                        }
+                    }
+                    ProposalDependency::QuorumProposal => {
+                        if let HotShotEvent::QuorumProposalValidated(proposal) = event {
+                            proposal.view_number
+                        } else {
+                            return false;
+                        }
+                    }
+                    ProposalDependency::PayloadAndMetadata => {
+                        if let HotShotEvent::QuorumProposalPayloadAndMetadataValidated(
+                            _payload_commitment,
+                            _metadata,
+                            view,
+                        ) = event
+                        {
+                            *view
+                        } else {
+                            return false;
+                        }
+                    }
+                };
+                event_view == view_number
+            }),
+        )
+    }
+
+    /// Create and store an [`AndDependency`] combining [`EventDependency`]s associated with the
+    /// given view number if it doesn't exist.
+    fn create_dependency_task_if_new(
+        &mut self,
+        view_number: TYPES::Time,
+        event_receiver: Receiver<Arc<HotShotEvent<TYPES>>>,
+        event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+        event: Arc<HotShotEvent<TYPES>>,
```

Comment:
Nit: Can we add a comment for `event`? Without comment, it seems that this function will create a task specifically for this event, but it actually will create a task that marks this event as completed.